cpp/include/cudf/io/datasource.hpp

```diff
@@ -314,6 +362,13 @@ class datasource {
   };
 };
 
+/**
+ * @brief Factory to construct a datasource object from a container.
+ *
+ * @tparam Container Type of the container to construct the datasource from
+ * @param data_owner The container to construct the datasource from (ownership is transferred)
+ * @return Constructed datasource object
```

Comment:
```suggestion
 * @brief Factory to construct a datasource buffer object from a container.
 *
 * @tparam Container Type of the container to construct the buffer from
 * @param data_owner The container to construct the buffer from (ownership is transferred)
 * @return Constructed buffer object
```

Reply:
This docstring is copied from the declaration part of the code in same file. doxygen throws warning due to a limitation of template function.
shall we move the code definition to the declaration itself?

Reply:
You can move the definition to the declaration, just need to forward declare `owning_buffer`. I think it's also possible to `copydoc` and avoid doc duplication that way.

Reply:
I tried `copydoc`, doxygen complains recursive reference.

Reply:
done.